File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return None

        tortoise = head
        hare = head

        while hare and hare.next:
            tortoise  = tortoise.next
            hare = hare.next.next


            if tortoise == hare:
                break
        else:
            return None


        tortoise = head
        
        while tortoise != hare :
            tortoise = tortoise.next
            hare = hare.next
        
        return tortoise
            

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head == null || head.next == null){
            return null;
        }
        ListNode slow = head;
        ListNode fast = head;
        boolean hasLoop = false;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast){
                hasLoop = true;
                break;
            }
        }

        if(!hasLoop){
            return null;
        }
         
       
        slow = head;
        while(slow != fast){
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}